core-api/src/test/java/com/optimizely/ab/event/internal/EventFactoryTest.java

```diff
@@ -249,6 +247,113 @@ public void createImpressionEventIgnoresUnknownAttributes() throws Exception {
+        boolean validBoolAttribute = true;
+        String validStringAttribute = "grayfindor";
+
+        HashMap<String, Object> attributes = new HashMap<>();
+        attributes.put(doubleAttribute.getKey(), validDoubleAttribute);
+        attributes.put(integerAttribute.getKey(), validIntegerAttribute);
+        attributes.put(boolAttribute.getKey(), validBoolAttribute);
+        attributes.put(stringAttribute.getKey(), validStringAttribute);
+
+        LogEvent impressionEvent =
+                factory.createImpressionEvent(projectConfig, activatedExperiment, bucketedVariation, "userId",
+                        attributes);
+
+        EventBatch impression = gson.fromJson(impressionEvent.getBody(), EventBatch.class);
+
+
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(0).getKey(), boolAttribute.getKey());
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(0).getValue(), validBoolAttribute);
+
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(1).getKey(), doubleAttribute.getKey());
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(1).getValue(), validDoubleAttribute);
+
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(2).getKey(), integerAttribute.getKey());
+        assertEquals((int) ((double) impression.getVisitors().get(0).getAttributes().get(2).getValue()), validIntegerAttribute);
+
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(3).getKey(), stringAttribute.getKey());
+        assertEquals(impression.getVisitors().get(0).getAttributes().get(3).getValue(), validStringAttribute);
+
+    }
+
+    /**
+     * Verify that passing through an null value attribute causes that attribute to be ignored, rather than
+     * causing an exception to be thrown.
+     */
+    @Test
+    public void createImpressionEventIgnoresNullAttributes() {
+        // use the "valid" project config and its associated experiment, variation, and attributes
+        ProjectConfig projectConfig = validProjectConfig;
+        Experiment activatedExperiment = projectConfig.getExperiments().get(0);
+        Variation bucketedVariation = activatedExperiment.getVariations().get(0);
+        Attribute attribute = validProjectConfig.getAttributes().get(0);
+
+        LogEvent impressionEvent =
+                factory.createImpressionEvent(projectConfig, activatedExperiment, bucketedVariation, "userId",
+                        Collections.singletonMap(attribute.getKey(), null));
+
+        EventBatch impression = gson.fromJson(impressionEvent.getBody(), EventBatch.class);
+
+        // verify that no Feature is created for attribute.getKey() -> null
+        for (com.optimizely.ab.event.internal.payload.Attribute feature : impression.getVisitors().get(0).getAttributes()) {
+            assertNotSame(feature.getKey(), attribute.getKey());
+            assertNotSame(feature.getValue(), null);
+        }
+    }
+
```

Comment:
Should we test attribute selection for _conversion_ events too?

BTW I think we could combine `createImpressionEventIgnoresInvalidAttributes`, `createImpressionEventWithIntegerDecimalBoolAndStringAttributes `, `createImpressionEventIgnoresNullAttributes` - that'd reduce complexity while also confirming that we merely ignore invalid attributes rather than dropping _all_ attributes when one or two are invalid.

Reply:
Lets leave these Unit Tests. I added one more unit test which is making ConversionEvent and in that unit test we are passing valid and invalid attributes in same attributes map, we are verifying that only valid attributes are getting passed and others are getting ignored. I think that will cover the condition you are asking to verify. So is that ok?

Reply:
Okay, that works too. Thanks!